libcalico-go/lib/set/adaptive.go

```diff
@@ -0,0 +1,351 @@
+		// We directly store a pointer to the item, which can also be seen
+		// as storing a single-element array.
+		a.p = unsafe.Pointer(&item)
+		a.size = 1
+	default:
+		// Sizes 1, 2, ..., adaptiveSetArrayLimit. Stored in an array.
+
+		// Make slice pointing at backing array with correct length/capacity.
+		tSlice := a.loadSliceFromPointer()
+
+		// First scan to see if the item is already present.
+		for _, t := range tSlice {
+			if t == item {
+				// The element is already in the set.
+				return
+			}
+		}
+
+		// If we get here, need to add the element to the set.
+		if a.size < adaptiveSetArrayLimit {
+			// Set is still small enough to use a slice.
+			if len(tSlice) == cap(tSlice) {
+				// Need to grow the slice; we do it manually so we can control
+				// the capacity and keep it in sync with arrCapForSize (allowing
+				// us to avoid storing it).
+				tSlice = growSliceCap(tSlice, arrCapForSize[a.size+1])
+				a.p = unsafe.Pointer(&tSlice[0])
+			}
+			tSlice = append(tSlice, item)
+			a.size++
+			return
+		}
+
+		// Too many items to store in a slice, upgrade to a map.
+		m := make(map[T]v, a.size+1)
+		for _, t := range tSlice {
+			m[t] = emptyValue
+		}
+		m[item] = emptyValue
+		a.p = unsafe.Pointer(&m)
+		a.size = sizeStoredInMap
+	case sizeStoredInMap:
+		m := a.loadMapFromPointer()
+		m[item] = emptyValue
+	}
+}
+
+func growSliceCap[T any](in []T, newCap int) []T {
+	out := make([]T, len(in), newCap)
+	copy(out, in)
+	return out
+}
+
+func (a *Adaptive[T]) AddAll(itemArray []T) {
+	for _, v := range itemArray {
+		a.Add(v)
+	}
+}
+
+func (a *Adaptive[T]) AddSet(other Set[T]) {
+	other.Iter(func(item T) error {
+		a.Add(item)
+		return nil
+	})
+}
+
+func (a *Adaptive[T]) Discard(item T) {
+	switch a.size {
+	case 0:
+		// Map is empty, nothing to discard.
+		return
+	case 1:
+		// Set has only one item, the pointer will point to that single item,
+		// which is equivalent to a single-entry array.
+		theOne := (*T)(a.p)
+		if *theOne == item {
+			a.p = nil
+			a.size = 0
+		}
+	default:
+		// Handles sizes 2, 3, ..., adaptiveSetArrayLimit. Stored in an array.
+
+		// Make slice pointing at backing array with correct length/capacity.
+		tSlice := a.loadSliceFromPointer()
+
+		// Scan the slice to see if the item is present.
+		for i, t := range tSlice {
+			if t == item {
+				// Found the element to remove.
+				newSize := a.size - 1
+				newCap := arrCapForSize[newSize]
+				if newCap < cap(tSlice) {
+					// Downgrade to a smaller array.
+					updatedSlice := make([]T, arrCapForSize[newSize])
+					// Copy the elements before and after the removed element.
+					copy(updatedSlice, tSlice[:i])
+					copy(updatedSlice[i:], tSlice[i+1:])
+					a.p = unsafe.Pointer(&updatedSlice[0])
+				} else {
+					// Keep the same slice.  Swap the last element into the
+					// removed element's slot.
+					tSlice[i] = tSlice[newSize]
+					var zeroT T
+					tSlice[newSize] = zeroT
+				}
+				a.size = newSize
+				return
```

Comment:
If we're aiming to be fully non-hysteretical with the underlying storage representation, I think we're missing the handling for reducing from size 2 to size 1.